### ISSUE TITLE
KeyError with joblib and sklearn cross_validate

### ISSUE THREAD

**Issue by ogrisel** [joblib/joblib#852]
This was originally reported as https://github.com/dask/distributed/issues/2532

I tried be reproduce it on joblib master with scikit-learn 0.20.2 as follows.

```python
import os
os.environ['SKLEARN_SITE_JOBLIB'] = "1"
from dask.distributed import Client
from sklearn import datasets, linear_model
from sklearn.model_selection import cross_validate
import joblib


client = Client(processes=False)
joblib.parallel_backend('dask')

diabetes = datasets.load_diabetes()
X = diabetes.data[:150]
y = diabetes.target[:150]
model = linear_model.LinearRegression()

cv_results = cross_validate(model, X, y, cv=10, return_train_score=False,
                            verbose=100)
```

However this seem to freeze without reporting the original error. Instead when I interrupt with ctrl-c I get:

```
[Parallel(n_jobs=-1)]: Using backend DaskDistributedBackend with 4 concurrent workers.
[CV]  ................................................................
[CV] .................................... , score=0.587, total=   0.0s
[Parallel(n_jobs=-1)]: Done   1 tasks      | elapsed:    0.0s
^CTraceback (most recent call last):
  File "/home/ogrisel/code/joblib/joblib/_dask.py", line 223, in maybe_to_futures
    f = call_data_futures[arg]
  File "/home/ogrisel/code/joblib/joblib/_dask.py", line 56, in __getitem__
    ref, val = self._data[id(obj)]
KeyError: 140545475489024

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ogrisel/tmp/joblib_dask_freeze.py", line 18, in <module>
    verbose=100)
  File "/home/ogrisel/code/scikit-learn/sklearn/model_selection/_validation.py", line 231, in cross_validate
    for train, test in cv.split(X, y, groups))
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 924, in __call__
    while self.dispatch_one_batch(iterator):
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 759, in dispatch_one_batch
    self._dispatch(tasks)
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 716, in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
  File "/home/ogrisel/code/joblib/joblib/_dask.py", line 254, in apply_async
    func, args = self._to_func_args(func)
  File "/home/ogrisel/code/joblib/joblib/_dask.py", line 243, in _to_func_args
    args = list(maybe_to_futures(args))
  File "/home/ogrisel/code/joblib/joblib/_dask.py", line 231, in maybe_to_futures
    [f] = self.client.scatter([arg])
  File "/home/ogrisel/code/distributed/distributed/client.py", line 1875, in scatter
    asynchronous=asynchronous, hash=hash)
  File "/home/ogrisel/code/distributed/distributed/client.py", line 676, in sync
    return sync(self.loop, func, *args, **kwargs)
  File "/home/ogrisel/code/distributed/distributed/utils.py", line 275, in sync
    e.wait(10)
  File "/opt/python3.7/lib/python3.7/threading.py", line 552, in wait
    signaled = self._cond.wait(timeout)
  File "/opt/python3.7/lib/python3.7/threading.py", line 300, in wait
    gotit = waiter.acquire(True, timeout)
KeyboardInterrupt
```

**Comment by jjerphan** [comment]
When executing as a module, with appropriate execution flow for modules, i.e:
```python
import os
os.environ['SKLEARN_SITE_JOBLIB'] = "1"
from dask.distributed import Client
from sklearn import datasets, linear_model
from sklearn.model_selection import cross_validate
import joblib

if __name__ == "__main__":
    client = Client()
    joblib.parallel_backend('dask')
    diabetes = datasets.load_diabetes()
    X = diabetes.data[:150]
    y = diabetes.target[:150]
    lasso = linear_model.Lasso()
    cv_results = cross_validate(lasso, X, y, cv=15, return_train_score=False, verbose=100)
```

it sometimes generally hangs forever. I am inspecting this right now.

---

`jobllib` was installed as editable from the github repo, see the environment bellow.
```bash
$ uname -v
Darwin Kernel Version 18.2.0: Thu Dec 20 20:46:53 PST 2018; root:xnu-4903.241.1~1/RELEASE_X86_64
$ pip freeze
appnope==0.1.0
atomicwrites==1.3.0
attrs==18.2.0
backcall==0.1.0
bleach==3.1.0
Click==7.0
cloudpickle==0.8.0
dask==1.1.2
decorator==4.3.2
defusedxml==0.5.0
distributed==1.26.0
entrypoints==0.3
HeapDict==1.0.0
ipykernel==5.1.0
ipython==7.3.0
ipython-genutils==0.2.0
ipywidgets==7.4.2
jedi==0.13.3
Jinja2==2.10
-e git+git@github.com:jjerphan/joblib.git@a99273bb0fe1d77c5e5b1b5512c6a9fa1285a125#egg=joblib
jsonschema==3.0.0
jupyter==1.0.0
jupyter-client==5.2.4
jupyter-console==6.0.0
jupyter-core==4.4.0
locket==0.2.0
MarkupSafe==1.1.1
mistune==0.8.4
more-itertools==6.0.0
msgpack==0.6.1
nbconvert==5.4.1
nbformat==4.4.0
notebook==5.7.4
numpy==1.16.2
pandas==0.24.1
pandocfilters==1.4.2
parso==0.3.4
partd==0.3.9
pexpect==4.6.0
pickleshare==0.7.5
pluggy==0.9.0
prometheus-client==0.6.0
prompt-toolkit==2.0.9
psutil==5.5.1
ptyprocess==0.6.0
py==1.8.0
Pygments==2.3.1
pyrsistent==0.14.11
pytest==4.3.0
python-dateutil==2.8.0
pytz==2018.9
PyYAML==3.13
pyzmq==18.0.0
qtconsole==4.4.3
scikit-learn==0.20.2
scipy==1.2.1
Send2Trash==1.5.0
six==1.12.0
sklearn==0.0
sortedcontainers==2.1.0
tblib==1.3.2
terminado==0.8.1
testpath==0.4.2
toolz==0.9.0
tornado==5.1.1
traitlets==4.3.2
wcwidth==0.1.7
webencodings==0.5.1
widgetsnbextension==3.4.2
zict==0.1.3
```

**Comment by ogrisel** [comment]
Note that the problem disappears when we set cv=5 for instance.

**Comment by robbisg** [comment]
Hi,

I digged a little bit into this issue, since I'm interested in running sklearn on a kubernetes cluster, and I found that it could be a data scattering/size problem.
I ran the same snippet in the first post, but I modified the size of ```X```.
```
X = diabetes.data[:131]
y = diabetes.target[:131]
```
and also increasing ```cv=20``` the program didn't hang.

The size of ```X``` (131) is the upper limit, above 131 the program hangs.

**Comment by robbisg** [comment]
Hi,

I've tried to reduce the memory allocated for numpy arrays by casting arrays in ```cross_validate``` flowing through joblib to float16 and int8, and this solution avoids the program to hang.
Obviously this solution is useless, but maybe you can figure out how to fix it!

```python
with parallel:
     scores = parallel(
          delayed(_fit_and_score)(
                clone(estimator), np.float16(X), np.float16(y), scorers, np.int8(train), np.int8(test), 
                verbose, None, fit_params, return_train_score=return_train_score,
                return_times=True, return_estimator=return_estimator, 
                error_score=error_score)

          for train, test in cv.split(X, y, groups))
```

**Comment by julioasotodv** [comment]
Having the same issue with another problem (that does not involve sklearn). 

Using joblib.Parallel + distributed and passing large arrays as argument for joblib.delayed raises the same exception.

**Comment by jjerphan** [comment]
> If so, it may be a problem with Tornado + client.scatter in dask.distributed, where large arrays are tried to be scattered inside a coroutine (which is not allowed, since client.scatter returns a Future result).

@julioasotodv: can you explain why scattering large arrays in a coroutine is not allowed? Thanks :)

**Comment by pierreglaser** [comment]
I dumped the stacktrace using `faulthandler` and @ogrisel's script:
```python
from dask.distributed import Client
from sklearn import datasets, linear_model
from sklearn.model_selection import cross_validate
import faulthandler
import joblib

faulthandler.dump_traceback_later(timeout=10, exit=True)

client = Client(processes=False)

diabetes = datasets.load_diabetes()
X = diabetes.data[:150]
y = diabetes.target[:150]
model = linear_model.LinearRegression()

joblib.parallel_backend("dask")

cv_results = cross_validate(
    model, X, y, cv=20, return_train_score=False, verbose=100
)
```

Stacktrace:
```
Thread 0x00007f61fb873700 (most recent call first):
  File "/usr/lib/python3.7/threading.py", line 300 in wait
  File "/usr/lib/python3.7/queue.py", line 179 in get
  File "/home/pierreglaser/repos/distributed/distributed/threadpoolexecutor.py", line 53 in _worker
  File "/usr/lib/python3.7/threading.py", line 865 in run
  File "/usr/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/usr/lib/python3.7/threading.py", line 885 in _bootstrap

Thread 0x00007f61fb072700 (most recent call first):
  File "/home/pierreglaser/repos/distributed/distributed/profile.py", line 272 in _watch
  File "/usr/lib/python3.7/threading.py", line 865 in run
  File "/usr/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/usr/lib/python3.7/threading.py", line 885 in _bootstrap

Thread 0x00007f61fa871700 (most recent call first):
  File "/home/pierreglaser/repos/joblib_master/joblib/parallel.py", line 307 in __call__
  File "/home/pierreglaser/repos/joblib_master/joblib/_dask.py", line 265 in callback_wrapper
  File "/home/pierreglaser/.virtualenvs/joblib_dask_py37/lib/python3.7/site-packages/tornado/gen.py", line 748 in run
  File "/home/pierreglaser/.virtualenvs/joblib_dask_py37/lib/python3.7/site-packages/tornado/gen.py", line 787 in inner
  File "/home/pierreglaser/.virtualenvs/joblib_dask_py37/lib/python3.7/site-packages/tornado/ioloop.py", line 743 in _run_callback
  File "/home/pierreglaser/.virtualenvs/joblib_dask_py37/lib/python3.7/site-packages/tornado/ioloop.py", line 690 in <lambda>
  File "/usr/lib/python3.7/asyncio/events.py", line 88 in _run
  File "/usr/lib/python3.7/asyncio/base_events.py", line 1764 in _run_once
  File "/usr/lib/python3.7/asyncio/base_events.py", line 528 in run_forever
  File "/home/pierreglaser/.virtualenvs/joblib_dask_py37/lib/python3.7/site-packages/tornado/platform/asyncio.py", line 148 in start
  File "/home/pierreglaser/repos/distributed/distributed/utils.py", line 411 in run_loop
  File "/usr/lib/python3.7/threading.py", line 865 in run
  File "/usr/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/usr/lib/python3.7/threading.py", line 885 in _bootstrap

Thread 0x00007f6204a41740 (most recent call first):
  File "/usr/lib/python3.7/threading.py", line 300 in wait
  File "/usr/lib/python3.7/threading.py", line 552 in wait
  File "/home/pierreglaser/repos/distributed/distributed/utils.py", line 330 in sync
  File "/home/pierreglaser/repos/distributed/distributed/client.py", line 763 in sync
  File "/home/pierreglaser/repos/distributed/distributed/client.py", line 2030 in scatter
  File "/home/pierreglaser/repos/joblib_master/joblib/_dask.py", line 232 in maybe_to_futures
  File "/home/pierreglaser/repos/joblib_master/joblib/_dask.py", line 244 in _to_func_args
  File "/home/pierreglaser/repos/joblib_master/joblib/_dask.py", line 255 in apply_async
  File "/home/pierreglaser/repos/joblib_master/joblib/parallel.py", line 716 in _dispatch
  File "/home/pierreglaser/repos/joblib_master/joblib/parallel.py", line 760 in dispatch_one_batch
  File "/home/pierreglaser/repos/joblib_master/joblib/parallel.py", line 926 in __call__
  File "/home/pierreglaser/.virtualenvs/joblib_dask_py37/lib/python3.7/site-packages/sklearn/model_selection/_validation.py", line 231 in cross_validate
  File "test_hang_joblib.py", line 19 in <module>
````

**Comment by ogrisel** [comment]
So indeed there is a deadlock between the `joblib.Parallel._lock` that is locked in the main thread at line 704 when calling `_dispatch` and then calls into `distributed.utils.sync` while the thread of the tornado IOLoop is wating for the same  `joblib.Parallel._lock` in the dask `callback_wrapper`.

**Comment by ogrisel** [comment]
And even ignoring the issue with `joblib.Parallel._lock`; the design of the dask backend is broken as we should not call `self.client.scatter` with `self.client.asynchronous=False` when called in the `callback_wrapper` because it is itself called in the thread of the tornado `IOLoop`.

@mrocklin do you have any suggestion as to how to refactor the scattering code to hande the sync vs async call contexts?

BTW, the issue is now reproduced as a test in #910 yielding the following traceback of the deadlock in travis:

https://travis-ci.org/joblib/joblib/jobs/559438497

**Comment by mrocklin** [comment]
I've looked briefly through the code and testing errors and I'm still confused about why this code would be called in an asynchronous context in the first place.

My understanding is that joblib isn't used in an async context today.  Does this mean that the joblib calls are happening within some Dask async context somehow?

Our general approach within Dask to handle sync/async switching is to write everything in an async-friendly manner, and then have a bit of magic very close to the user API to sync-ify everything if necessary.  I'm not sure that that approach is appropriate for Joblib though.  My guess is that tracking down why this was running in an async context in the first place would be more valuable, though again I don't really have a good idea for what is going on here.

**Comment by pierreglaser** [comment]
Thanks for looking into it @mrocklin .

> I've looked briefly through the code and testing errors and I'm still confused about why this code would be called in an asynchronous context in the first place.

In `joblib` actally, most of the tasks to be executed are dispatched within async callbacks -- the `distributed` specific callbacks can be found here:
https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/_dask.py#L252-L267

The `callback` argument of `apply_async` is a function that will dispatch the next task (if any) of the original task iterator.

**Comment by mrocklin** [comment]
So if we're running into sync/async issues, that might imply that we're calling `scatter` from inside of one of these async or `gen.coroutine` functions.  Do we have a sense for how this might occur?

**Comment by ogrisel** [comment]
The `scatter` is called by the `apply_async` method. The  `apply_async` can be either called by the `joblib.Parallel.__call__` in the main thread or via the `callback` that is executed asynchronously in the `callback_wrapper` coroutine in the thread that runs the tornado event loop.

Basically `apply_async` calls are chained in joblib: once a task at completed it triggers the submission of the next (batch of) tasks via the callback.

**Comment by jjerphan** [comment]
> Do we have a sense for how this might occur?

I think yes: when a batch of tasks is completed, the callback that is executed is the `BatchCompletionCallback` that indirectly calls `scatter`.

The sequence of execution is as follows:
 - `callback_wrapper` (in the `Tornado's IO.Loop`)
https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/_dask.py#L259-L264
 - `BatchCompletionCallBack.__call__`:
https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/parallel.py#L300-L309
 - `Parallel.dispatch_next`:
https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/parallel.py#L723-L733
 - `Parallel.dispatch_one_batch`:
https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/parallel.py#L735-L760
- `Parallel._dispatch`:
https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/parallel.py#L698-L721
- `DaskDistributedBackEnd.apply_async`:
https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/_dask.py#L252-L255
- and `DaskDistributedBackEnd._to_func_args` that finally calls `client.scatter` via `maybe_to_futures` [when the arguments' size exceed a threshold](https://github.com/joblib/joblib/issues/852#L225) (line 225 to 235)
https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/_dask.py#L205-L250

I don't really understand the logic of this block. @julioasotodv and I explored this a bit on [`distributed/issues/2149`](https://github.com/dask/distributed/issues/2149#issuecomment-507797590) that we believe is related to this issue.

**Comment by jjerphan** [comment]
@ogrisel: as tasks are dispatched in `Parallel.__call__`, why do we need the `BatchCompletionCallBack` to call `dispatch_next` exactly?

**Comment by ogrisel** [comment]
> @ogrisel: as tasks are dispatched in Parallel.__call__, why do we need the BatchCompletionCallBack to call dispatch_next exactly?

Only a subset of the tasks are dispatched directly via `Parallel.__call__`. The remaining tasks in the task iterator are consumed via the callback. The goal is not to consume the iterator eagerly as joblib was designed such that the task could be read from the disk on the fly for instance and we would not have to pre-load everything in memory prior to start the dispatching.

**Comment by jjerphan** [comment]
> Only a subset of the tasks are dispatched directly via Parallel.__call__. The remaining tasks in the task iterator are consumed via the callback. The goal is not to consume the iterator eagerly as joblib was designed such that the task could be read from the disk on the fly for instance and we would not have to pre-load everything in memory prior to start the dispatching.

I see. How this was done is indeed explained in [the method's comments](https://github.com/joblib/joblib/blob/b883dc2de5fa45723589139a4d331b717a15c809/joblib/parallel.py#L907-L922) but I was trying to understand why. Thank you for the explanation!

**Comment by mrocklin** [comment]
> The apply_async can be either called by the joblib.Parallel.__call__ in the main thread or via the callback that is executed asynchronously in the callback_wrapper coroutine in the thread that runs the tornado event loop.

If `scatter` is called from within the Tornado event loop's thread then it will magically choose to run as a `coroutine`/`async def function`.  It can't run as a normal blocking function in this context because it needs the event loop to be active in handling communications to the scheduler and workers.

There are two options here:

1.  Coordinate things so that we never call `scatter` from the event loop, and never block on its finishing from within the event loop (which will be tricky).  
2.  Use async-def or coroutine code here.

Choosing option one will be tricky.  We really shouldn't be doing any blocking operations within the event loop, so even something like running scatter in a separate thread and blocking until it finished would probably not work.

**Comment by ogrisel** [comment]
Indeed this is what I thought. I wonder why we don't get a similar issue with the same issue with the call to `self.client.submit` but apparently always work as a synchronous call probably because it's expected to always be fast and therefore would not really benefit from an `asynchronous=True` option.

To implement solution 2. we would need to refactor all the functions called under `apply_async` to work either as syncrhonous functions or awaitable coroutines using the `self.client.sync` trick already used in `self.client.scatter`.  I will try to find the time to give it a try this afternoon or tomorrow. If anywayone else wants to try, please feel free to do it as well though :)

**Comment by mrocklin** [comment]
> I wonder why we don't get a similar issue with the same issue with the call to self.client.submit but apparently always work as a synchronous call probably because it's expected to always be fast and therefore would not really benefit from an asynchronous=True option.

Yes, that's correct.

**Comment by pierreglaser** [comment]
I implemented a possible fix in #914 for which the sklearn reproducer as well as the new joblib tests of #910 pass. @jjerphan @julioasotodv feel free to test this fix against your use-cases, feedback is welcomed :)

**Comment by jjerphan** [comment]
Thanks @pierreglaser ; I'll have a look at it ASAP (cc @samronsin).
